xds/src/main/java/io/grpc/xds/sds/SdsClient.java

```diff
@@ -0,0 +1,308 @@
+import io.envoyproxy.envoy.service.discovery.v2.SecretDiscoveryServiceGrpc.SecretDiscoveryServiceStub;
+import io.grpc.Internal;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.netty.NettyChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.EpollDomainSocketChannel;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.unix.DomainSocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * SDS client used by {@link SslContextProvider}s to get SDS updates from the SDS server. This is
+ * most likely a temporary implementation until merged with the XdsClient.
+ */
+// TODO(sanjaypujare): once XdsClientImpl is ready, merge with it and add retry logic
+@Internal
+final class SdsClient {
+  private static final Logger logger = Logger.getLogger(SdsClient.class.getName());
+  private static final String SECRET_TYPE_URL = "type.googleapis.com/envoy.api.v2.auth.Secret";
+
+  private final HashMap<String, HashSet<SecretWatcher>> watcherMap = new HashMap<>();
+
+  private static SdsClient instance;
+  private final ConfigSource configSource;
+  private final Node clientNode;
+  @VisibleForTesting String targetUri;
+  ManagedChannel channel;
+  private SecretDiscoveryServiceStub secretDiscoveryServiceStub;
+  private ResponseObserver responseObserver;
+  private StreamObserver<DiscoveryRequest> requestObserver;
+  private DiscoveryResponse lastResponse;
+
+  /**
+   * Starts resource discovery with SDS protocol. This method should be the first one to be called
+   * and should be called only once.
+   */
+  void start() {
+    if (channel == null) {
+      if (targetUri.startsWith("unix:")) {
+        EventLoopGroup elg = new EpollEventLoopGroup();
+        channel =
+            NettyChannelBuilder.forAddress(new DomainSocketAddress(targetUri.substring(5)))
+                .eventLoopGroup(elg)
+                .channelType(EpollDomainSocketChannel.class)
+                .build();
+      } else {
+        channel = NettyChannelBuilder.forTarget(targetUri).build();
+      }
+    }
+    secretDiscoveryServiceStub = SecretDiscoveryServiceGrpc.newStub(channel);
+    responseObserver = new ResponseObserver();
+    requestObserver = secretDiscoveryServiceStub.streamSecrets(responseObserver);
+  }
+
+  /** Stops resource discovery. No method in this class should be called after this point. */
+  void shutdown() {
+    requestObserver.onCompleted();
+    channel.shutdownNow();
+  }
+
+  /**
+   * Gets the singleton instance after constructing it as needed.
+   *
+   * @param configSource the {@link ConfigSource} pointing to the SDS server.
+   * @param node Node id of this client.
+   */
+  static synchronized SdsClient getInstance(ConfigSource configSource, Node node) {
```

Comment:
Another PR will have code calling this method. Basically the `ClientSslContextProviderFactory` or `ServerSslContextProviderFactory` will call this based on received SdsConfig (and the `ConfigSource` inside it. Because of the recent discovery that I need one SdsClient per resource name I will have to restructure this now.

Reply:
this code is gone